### PR TITLE
Electron todo cleanups

### PR DESCRIPTION
**Pull request by mfonville**
Some todos that were mentioned in the code for when Electron would be upgrade

@sadick254 should work, but I am not able to run any tests for this myself locally

### CODE REVIEW

File: spec/integration/helpers/start-atom.js [hunk 1]
```diff
@@ -59,9 +59,6 @@ const buildAtomClient = async (args, env) => {
   // So, for now, we'll just use a random port (we don't care about its value since we're not
   // connecting through it).
   // (inspired by https://github.com/electron/spectron/pull/361/commits/737db138bd8a6daaf80f9c2bff710ce4a5fff39b).
-  // TodoElectronIssue: Remove the whole remote-debugging-port param once we upgrade
-  // to Electron v5, since this was fixes there (see electron/electron#17800).
-  const randomPort = Math.floor(Math.random() * (9999 - 9000) + 9000);
```
Comment (DeeDeeG) [inline spec/integration/helpers/start-atom.js:64]: I think this entire comment is describing the need for the workaround. I think the whole comment can be deleted given that the workaround is apparently not needed anymore.

Comment (DeeDeeG) [reply spec/integration/helpers/start-atom.js:64]: (I wasn't able to highlight the whole comment, but I meant all eight lines of the comment.)

Comment (mfonville) [reply spec/integration/helpers/start-atom.js:64]: I believe the first part of the comment is still relevant. Because there is still a random port set-up for the debugger. The only thing is that the problem of with it being set to zero (which is the default value), which means to make it random, was fixed in the newer Electron releases.

Comment (DeeDeeG) [reply spec/integration/helpers/start-atom.js:64]: If you visit the PR mentioned, the discussion there is strictly based on solving the ChromeDriver issue this whole comment is talking about.

https://github.com/electron/electron/pull/17800

> Notes: Fixed issue where `chromedriver` would not connect correctly if you didn't provide a custom `remote-debugging-port`

Electron sets this random port for us, to satisfy ChromeDriver's rather arbitrary requirements. _(Knowing about this is not important to Atom developers, as long as ChromeDriver generally behaves itself. According to the comment, we don't actually use this port for anything, other than (one infers) letting ChromeDriver manage and use this port in a fully automated way.)_ So the behavior described in the comment lives in Electron now, not Atom's codebase. With the fix landed at Electron, all of this is just acting with the expected/default behaviors, which doesn't require a comment to describe it at this repo.

It's interesting to have an insight into what the code is doing, or historical context for why the code is the way it is, and more comments are usually better, but I think the convention is to not have comments for things that are working the usual/expected way, and for things that don't require special attention from devs.

In other words, this isn't actionable information anymore, so encountering this comment might mislead readers that there is something they have to do about this issue, or some care they need to take. In fact, it was effectively a bug in upstream Electron, which at one point Atom had to work around, but which has since been fixed. If the workaround is deleted, the comment justifying it logically should be cleaned up along with the workaround.

To the extent that it remains relevant, the comment "lives on" in the commit history of this file. _(If delving into the past, any place folks encounter this workaround back in the commit history, the comment will be preserved there along with it.)_

Comment (mfonville) [reply spec/integration/helpers/start-atom.js:64]: You are right. Since I am just cleaning it up, I thought that the first part might still be relevant for Atom devs too, as I thought using ChromeDriver was non-trivial. But if it is, than it is indeed not necessary information. I will update the PR

File: spec/integration/helpers/start-atom.js [hunk 2]
```diff
@@ -80,7 +70,6 @@ const buildAtomClient = async (args, env) => {
           'dev',
           'safe',
           `user-data-dir=${userDataDir}`,
```
Comment (DeeDeeG) [inline spec/integration/helpers/start-atom.js:72]: [The linter pointed this out](https://dev.azure.com/DeeDeeG/b/_build/results?buildId=935&view=logs&j=3a559e2a-952e-58d2-b8db-2e604a9266d7&t=6f845a59-a86d-5907-2b45-02c2fb4984eb&l=14): No comma needed, because this is the last item in the list.

Other than this failing linting due to the extra comma, CI passes if applying these changes on top of `master`:

https://dev.azure.com/DeeDeeG/b/_build/results?buildId=935

**Comment by DeeDeeG** [comment]
I addressed the remaining `todoElectronIssue` comments. CI is passing (other than two linting complaints that I missed).

https://github.com/DeeDeeG/atom/commits/electron-todo-cleanups-6

@mfonville would you want these commits as a PR into your branch/fork?

**Comment by mfonville** [comment]
@DeeDeeG yes please!

**Comment by DeeDeeG** [comment]
Update: Reworking one of the fixes and testing again.

**Review by aminya — CHANGES_REQUESTED** [review]
What is the result of CI when you remove these?

Could you add a link to Azure for a sample run?

File: src/native-compile-cache.js [hunk 3]
```diff
@@ -38,12 +38,11 @@ class NativeCompileCache {
...
   runInThisContext(code, filename) {
-    // TodoElectronIssue:  produceCachedData is deprecated after Node 10.6, so we'll
-    // will need to update this for Electron v4 to use script.createCachedData().
-    const script = new vm.Script(code, { filename, produceCachedData: true });
+    const script = new vm.Script(code, filename);
+    const cachedData = script.createCachedData();
     return {
       result: script.runInThisContext(),
-      cacheBuffer: script.cachedDataProduced ? script.cachedData : null
+      cacheBuffer: typeof cachedData !== 'undefined' ? cachedData : null
```
Comment (aminya) [inline src/native-compile-cache.js:45]:
```suggestion
      cacheBuffer: cachedData !== undefined ? cachedData : null
```

File: src/main-process/atom-application.js [hunk 4]
```diff
@@ -278,23 +268,18 @@ module.exports = class AtomApplication extends EventEmitter {
...
-    // which is really slow on Windows machines.
-    // (TodoElectronIssue: This got fixed in electron v3: https://github.com/electron/electron/issues/2073).
     let socketServerPromise;
     if (options.test || options.benchmark || options.benchmarkTest) {
       socketServerPromise = Promise.resolve();
     } else {
       socketServerPromise = this.listenForArgumentsFromNewProcess();
     }
 
+    await socketServerPromise;
```
Comment (aminya) [inline src/main-process/atom-application.js:278]: What is the benefit of awaiting this earlier? If it is not needed why we should do it?

Comment (mfonville) [reply src/main-process/atom-application.js:278]: @DeeDeeG

Comment (DeeDeeG) [reply src/main-process/atom-application.js:278]: > Benchmarking the await would help as well.

If someone other than me could handle this, that would be great. I'm all for learning new things, but someone who has benchmarked JS before might better be able to tell whether the results look good or bad.

> What is the benefit of awaiting this earlier?

Supposedly it prevents a race condition. This listener listens for requests to open new windows, and somehow transfers these requests/their metadata to the main window or process. (According to the PRs I link below). And it really isn't much slower, maybe under ~20ms usually? Or not slower at all? Hard to tell, honestly, since startups are so variable already. But I used the builtin [`timecop`](https://github.com/atom/timecop) package, and it's similar with or without these tweaks. I am relying on the commentary of Atom devs in the pull requests that implemented the workaround in the first place. See discussion here: https://github.com/atom/atom/pull/19242 & https://github.com/atom/atom/pull/19246 -- this code was tweaked a bit more in https://github.com/atom/atom/pull/19354 & https://github.com/atom/atom/pull/19455.

**Comment by DeeDeeG** [comment]
In general, please see my notes for my changes in this PR over at @mfonville's fork: https://github.com/mfonville/atom/pull/1

My thoughts are given for each change, and I linked to the PRs where the trickier of these workarounds was introduced/tweaked (the `socketServerPromise` / `listenForArgumentsFromNewProcess` change).

File: src/native-compile-cache.js (hunk 3 above)
Comment (DeeDeeG) [reply src/native-compile-cache.js]: You can't directly compare the value of something you expect to be undefined, or you will get this error:
`Uncaught ReferenceError: cachedData is not defined`

`typeof` can check if something is undefined without causing a `ReferenceError`. I don't want to start causing Errors and having to handle them correctly. Best to use `typeof` here.

**Comment by DeeDeeG** [comment]
> What is the result of CI when you remove these?
> 
> Could you add a link to Azure for a sample run?

https://dev.azure.com/DeeDeeG/b/_build/results?buildId=953

This run has just this branch's changes rebased on top of `master`, because the Electron 9 branch isn't building right now, unrelatedly to this PR.

---

I also did a bunch of runs with the startup listener thing enabled during testing. See here: [CI link](https://dev.azure.com/DeeDeeG/b/_build?definitionId=17&branchFilter=113) (Runs with commit hash d13cc88.)

---

I see this error in most (but not all) runs, so that needs looking into:

```
##[error] Tests for core-render-process in D:/a/1/s/spec/pane-spec.js. failed.

Pane
  pending state
    it Pane pending state does not open file in pending state by default
      timeout: timed out after 120000 msec waiting for spec promise to resolve
```

File: src/native-compile-cache.js (hunk 3 above)
Comment (mfonville) [reply src/native-compile-cache.js]: ok will revert it again; thanks for the explanation

**Comment by mfonville** [comment]
> I also did a bunch of runs with the startup listener thing enabled during testing. See here: [CI link](https://dev.azure.com/DeeDeeG/b/_build?definitionId=17&branchFilter=113) (Runs with commit hash [d13cc88](https://github.com/atom/atom/commit/d13cc8802404baecc8d9ec7d629a29866e235d94).)
> 
> I see this timeout in one or two (but not all) runs, so that needs looking into:
> 
> ```
> ##[error] Tests for core-render-process in D:/a/1/s/spec/pane-spec.js. failed.
> 
> Pane
>   pending state
>     it Pane pending state does not open file in pending state by default
>       timeout: timed out after 120000 msec waiting for spec promise to resolve
> ```

Do those timeouts also happen on `master`? If not, should we maybe bisect which of the changes triggers this?

**Comment by DeeDeeG** [comment]
> Do those timeouts also happen on `master`?

I don't recall seeing that on `master` in Azure Pipelines. On my slow MacBook Pro with a spinning disk drive, I get 4-7 errors on the `pane-spec.js` tests, consistently, even on `master`. On a fast Windows machine with an SSD that I borrowed, all tests passed. Which I would infer means there is a (potentially minor) slowdown somewhere on this branch impacting the pane tests. Or just Azure Pipelines being weird.

> If not, should we maybe bisect which of the changes triggers this?

Yeah, bisecting would be a good idea. I've been testing this a lot, so maybe I'll come back to it, but I can't commit to it in the next day or so, including the holiday.

**Review by aminya — CHANGES_REQUESTED** [review]
Could you change the target of this PR to the master branch and revert anything that requires Electron 9? 

These changes require passing the tests, and so it is better to merge them to the master branch instead. Merging these into Electron 9 branch may introduce unwanted bugs that are hard to debug.

File: resources/linux/debian/control.in [hunk 5]
```diff
@@ -1,6 +1,6 @@
 Package: <%= appFileName %>
 Version: <%= version %>
-Depends: git, libgconf-2-4 (>= 3.2.5), libgtk-3-0 (>= 3.9.10), libgcrypt20, libnotify4, libxtst6, libnss3 (>= 2:3.22), libglib2.0-bin | kde-cli-tools | kde-runtime, xdg-utils, libx11-xcb1, libxss1, libxkbfile1, libcurl3 | libcurl4
```
Comment (aminya) [inline resources/linux/debian/control.in:3]: Move this to another PR

File: script/vsts/platforms/templates/preparation.yml [hunk 6]
```diff
@@ -4,7 +4,7 @@ steps:
   - script: |
       sudo apt-get update
       sudo apt-get install -y wget software-properties-common
-      sudo apt-get install -y build-essential ca-certificates xvfb fakeroot git libsecret-1-dev rpm libx11-dev libxkbfile-dev xz-utils xorriso zsync libxss1 libgconf2-4 libgtk-3-0 libasound2 libicu-dev
+      sudo apt-get install -y build-essential ca-certificates xvfb fakeroot git libsecret-1-dev rpm libx11-dev libxkbfile-dev xz-utils xorriso zsync libxss1 libgtk-3-0 libasound2 libicu-dev
```
Comment (aminya) [inline script/vsts/platforms/templates/preparation.yml:7]: Move this to another PR

Comment (mfonville) [reply script/vsts/platforms/templates/preparation.yml:7]: Yeah, actually they are already merged in in the electron-9 branch, they are leftovers of an old PR.

File: resources/linux/debian/control.in (hunk 5 above)
Comment (mfonville) [reply resources/linux/debian/control.in:3]: Yeah, actually they are already merged in in the electron-9 branch, they are leftovers of an old PR.

**Comment by mfonville** [comment]
I will rebase on master as suggested

**Comment by mfonville** [comment]
@darangi I think this should be ready for a review by the Atom team

**Comment by DeeDeeG** [comment]
I did manage to get some benchmarks for `socketServerPromise` on all three OSes, comparing before/after the changes from this PR. I have to get an opportunity to grab the data off the differrent machines and make some sense of it. But wanted to mention that in case this would get reviewed/merged before I got a chance to post the numbers.

**Comment by joseph5119** [comment]
The connection is secure,  now pass phase and future to hand back into present

File: src/main-process/atom-application.js (hunk 4 above)
Comment (Aerijo) [reply src/main-process/atom-application.js:278]: There is a race condition regardless. If you start two instances concurrently with `--inspect-brk=<port>` to pause them you can walk through and have both think they are establishing the socket. Both will check for the socket secret file (determined by Atom home + version; but none exists), both will generate independent secrets (just a random value), both will write their secret to the same socket secret file (only the last to write will persist), and both will establish servers listening to the named pipe corresponding to their secret.

An instance has already decided to keep going once this point is reached, and no socket related behaviour needs to happen strictly after the socket is established, so in terms of correctness it does not matter where the socket promise is awaited. AFAICT, the only reason to await it at all is to measure it's impact. The promise it becomes a part of is [not used](https://github.com/atom/atom/blob/72709dcaa5a6caf4fb7ea5426ffbaceba6ea1083/src/main-process/start.js#L124). 

(the use of the socket secret file is for new instances to find it, read the contained secret, deterministically convert this secret into the name of a named pipe for inter process communication, and inform the existing instance monitoring this pipe of the arguments to this new one).

Comment (DeeDeeG) [reply src/main-process/atom-application.js:278]: Thanks for that info.

With the debugger off, I think there is some (small) real-world impact. When I tried spamming "atom" in `cmd.exe` many times in a second, with Atom not started up yet, I informally was able to reproduce that the Atom windows ended up in fewer independent processes with this change. Would have to try again and confirm before I was totally confident in this result being meaningful though.

I think the impact is more profound with a very slow/resources-tied-up machine that really chugs during startup. (And maybe moreso for "cold starts"? Atom is dramatically slower to start when it hasn't been cached in RAM yet, in my experience.)

**Comment by DeeDeeG** [comment]
Regarding benchmarking, I'm reluctant to just post a dump the numbers, because the story they tell doesn't make much sense (to me) without some inferring and educated guessing to fill in the gaps -- I have "exact data" for how long `socketServerPromise` takes, but perhaps more importantly, startup time to a usable editor window (measured by `timecop` or with a stopwatch) seems not to be measurably affected.

Caveats and commentary:

Measuring before and after `await socketServerPromise` generally appears to "take longer" (~50ms vs ~14ms) when moved up sooner. (However, on one of my Windows machines, the promise takes longer to resolve on `master` branch, without these changes.) The reason I don't think the apparently longer `await socketServerPromise` duration is going to hurt end-users is that it doesn't appear to result in an overall slower startup,  in terms of arriving at a visible window or to a usable editor. I think moving this up sooner may make it concurrently fight for resources with some other activity and _appear_ to take longer (on three out of four of my test machines) in the benchmark data. Whereas all the work has to get done anyway in order to start the app properly. And `timecop` doesn't show a measurably (through the noise) faster or slower startup either way.

I don't see why the same amount of work in the `socketServerPromise` would take more or less time based on whether it's moved up earlier or not, but it does come back with a different duration in the benchmark data. That's why I think it's fighting for resources or doing some waiting for other async work or disk access.

In other words, I can get data that make this appear to take longer when it's moved up earlier, but I think that's not really what's going on. (And the opposite effect appears in the data for one of my Windows machines??) In _other_ other words, I don't think this benchmark data is useful without context, and the context is arguably more important than the one piece of hard data that is the `await socketServerPromise` duration. But you may come to your own conclusions, as I am posting the data.

---

General observations:

`await socketServerPromise` "takes much longer" on cold startups. (That is, any time Atom is not cached in RAM). My guess: Probably waiting for disk activity and other async work going on at the same time.

When measuring before and after `await socketServerPromise`, _the results show a longer duration if the `await` is moved up earlier_, generally speaking. (But the opposite is true on one of my (borrowed) Windows machines.) I suspect this is because it is happening at a busier time and contending with other async work. But I'm not really sure.

All of the _other_ `TodoElectronIssue` fixes together do not appear to measurably affect the time to `await socketServerPromise`. Furthermore: All of the fixes combined (including moving `await socketServerPromise` up earlier) do not appear to measurably affect the time to get a usable editor, on any OS.

---

Benchmarking results:

_(Please do not skip my comments above, they add important context to these numbers. Overall startup time is not noticeably affected; Total startup time is very variable from run to run, and this variation is much larger than the time that `await SocketServerPromise` takes, so determining if this affects startup time positively or negatively strictly from this data isn't doable. In my view, the goal is fast startup for users, and modern correct behavior under the hood, not performance in micro-benchmarks.):_

<details><summary>Expand if you have read my comments above.</summary>

On Linux/macOS, on somewhat old but still strong computers for ~8-10 years ago (HDD),  `await SocketServerPromise` is either near-instantaneous (~0.05 to ~0.09 ms on Linux/macOS `master` branch) or about 45-50ms (macOS) or 10-50 ms (Linux) if moved up.

On Windows, on a system with a slow CPU and an HDD spinning disk drive, `await socketServerPromise` took about ~0.12-~0.20 ms on `master` branch. When moving up the `await socketServerPromise`, it took about ~33-46 ms. 

On Windows again, on a system with an SSD and a recent AMD Ryzen desktop APU, `await socketServerPromise` took (occasionally) about ~0.05 ms OR (rarely) ~3-5 ms OR (usually) ~90-100 ms on `master` branch, with no apparent pattern or cause (that I could tell) for the variation.  With the `await` moved up earlier, `await socketServerPromise` took more-consistently about ~13-17 ms.

</details>

**Comment by DeeDeeG** [comment]
Regarding the `pane-spec` timeouts mentioned earlier in this thread, I've been seeing those elsewhere. So I'm going to go ahead and say I don't think it has to do with this PR.

**Review by sadick254 — APPROVED** [review]
This looks good to me. Thank you @mfonville for the cleanup.
